pkg/src/src/main/scala/edu/berkeley/cs/amplab/sparkr/SparkRRunner.scala

```diff
@@ -0,0 +1,97 @@
+package edu.berkeley.cs.amplab.sparkr
+
+import java.io._
+import java.net.URI
+import java.util.concurrent.Semaphore
+
+import scala.collection.mutable.ArrayBuffer
+import scala.collection.JavaConversions._
+
+import org.apache.hadoop.fs.Path
+
+/**
+ * Main class used to launch SparkR applications using spark-submit. It executes R as a
+ * subprocess and then has it connect back to the JVM to access system properties etc.
+ */
+object SparkRRunner {
+  def main(args: Array[String]) {
+    val rFile = args(0)
+
+    val otherArgs = args.slice(1, args.length)
+
+    // TODO: Can we get this from SparkConf ?
+    val sparkRBackendPort = sys.env.getOrElse("SPARKR_BACKEND_PORT", "12345").toInt
+    val rCommand = "Rscript"
+
+    // Check if the file path exists.
+    // If not, change directory to current working directory for YARN cluster mode
+    val rF = new File(rFile)
+    val rFileNormalized = if (!rF.exists()) {
+      new Path(rFile).getName
+    } else {
+      rFile
+    }
+
+
+    // Launch a SparkR backend server for the R process to connect to; this will let it see our
+    // Java system properties etc.
+    val sparkRBackend = new SparkRBackend()
+    val sparkRBackendThread = new Thread() {
+      val finishedInit = new Semaphore(1)
+      finishedInit.acquire()
```

Comment:
It seems to me we should

* initialize the semaphore to have 0 permit
* remove the acquire inside here (but keep the one in the caller thread)

The current setup seems to have a risk of the caller thread acquiring first, then this constructor will deadlock?